zebra-chain/src/history_tree.rs

```diff
@@ -0,0 +1,245 @@
+//! History tree (Merkle mountain range) structure that contains information about
+//! the block history as specified in ZIP-221.
+
+use std::{
+    collections::{BTreeMap, HashSet},
+    io,
+    sync::Arc,
+};
+
+use thiserror::Error;
+
+use crate::{
+    block::{Block, ChainHistoryMmrRootHash, Height},
+    orchard,
+    parameters::{Network, NetworkUpgrade},
+    primitives::zcash_history::{Entry, Tree},
+    sapling,
+};
+
+/// An error describing why a history tree operation failed.
+#[derive(Debug, Error)]
+#[non_exhaustive]
+#[allow(missing_docs)]
+pub enum HistoryTreeError {
+    #[error("zcash_history error: {inner:?}")]
+    #[non_exhaustive]
+    InnerError { inner: zcash_history::Error },
+
+    #[error("I/O error")]
+    IOError(#[from] io::Error),
+}
+
+/// History tree structure.
+pub struct HistoryTree {
+    network: Network,
+    network_upgrade: NetworkUpgrade,
+    /// Merkle mountain range tree.
+    /// This is a "runtime" structure used to add / remove nodes, and it's not
+    /// persistent.
+    inner: Tree,
+    /// The number of nodes in the tree.
+    size: u32,
+    /// The peaks of the tree, indexed by their position in the array representation
+    /// of the tree. This can be persisted to save the tree.
+    peaks: BTreeMap<u32, Entry>,
+    /// The height of the most recent block added to the tree.
+    current_height: Height,
+}
+
+impl HistoryTree {
+    /// Create a new history tree with a single block.
+    pub fn new_from_block(
```

Comment:
Per the [Rust API Guidelines](https://rust-lang.github.io/api-guidelines/naming.html), I would name this conversion constructor `from_block`:

```suggestion
    pub fn from_block(
```

Reply:
Ah good suggestion, changed in 97c4110